### ISSUE TITLE
[Haskell] stack only setup

### ISSUE THREAD

**Issue by boj** [syl20bnr/spacemacs#2142]
I recently removed everything Haskell related and rebuilt my environment using only `stack`.  It currently uses lts-2.15 with ghc-7.8.4.

Unfortunately this ends up with a handful of my imports in spacemacs saying "Could not find module 'X'.  Use -v to see a list of the files searched for."

The following globally installed to `~/.local/bin`, which is also added to my $PATH:

    stack install stylish-haskell hlint ghc-mod hasktags

And set the following in my `.spacemacs`:

    (add-to-list 'exec-path "~/.local/bin")

I suppose it's not finding the application-specific libraries I have installed with stack in my project's working directory?

Clearly this is new beta software and isn't expected to work with spacemacs out of the box.  I was just curious if anyone has tips on how to get it to work, or if there is a fix somewhere upstream in the works?

**Comment by kvelicka** [comment]
I think you'll need to provide a little more information to get help with this. Some things you're not telling us:

1. What modules were not found?
2. What are you trying to do when you get the warnings (open a file, open a project etc)?
3. Are the modules that spacemacs can't find related to its layers (i.e. the stylish-haskell, hlint, etc that you mentioned) or to the project you're trying to open/compile/run/etc (again, you didn't specify what you were trying to do).

I'd like for spacemacs to have better Haskell support but that requires giving as comprehensive bug reports as possible.

**Comment by boj** [comment]
@kvelicka Happy to help.

What I am seeing is the following:

![screen shot 2015-06-29 at 10 14 29 pm](https://cloud.githubusercontent.com/assets/50839/8408603/59ce9a5e-1eac-11e5-8f39-f5b384a2bc4b.png)

Basically when I put my cursor over an import, it tells me some of them are not found.

There isn't an actual spacemacs error, more than likely the haskell-mode layer stuff.  I'm just not sure how to interpret it, or how to debug it.

**Comment by kvelicka** [comment]
Right, thanks for elaborating. One more thing - are the modules, say Conduit in the screenshot you just attached, installed globally or in a sandbox?

**Comment by hakuch** [comment]
I'm also interested in getting stack working seamlessly with Spacemacs and it's Haskell support.

I'm using the `master` branch now, but I can test changes on `develop`.

Placing [this](https://github.com/commercialhaskell/stack/blob/master/.dir-locals.el) file in my `~` directory (minus the formatting stuff) means that I can load files into the REPL, but much of the other functionality (company completion, type inspection, e.t.c.) seems to be missing.

I should also mention that ghc-mod doesn't seem to be loaded, and error marking isn't working either.

I'm not exactly sure where to start as far as debugging, but i'm happy to help in any way that I can.

**Comment by kvelicka** [comment]
Regarding `ghc-mod`, I can't even get it to compile on 7.10.1. I'm not using `stack` but I feel that most of the problems raised in this thread aren't necessarily stack-specific.

**Comment by boj** [comment]
@kvelicka I have installed modules like Conduit using `stack install <package>` from my working directory, which if I understand correctly adds them to a project specific sandbox database?

Another thing which has been happening is that my editor hangs with `Initializing...` when I try to open one of my Haskell files.  I'm not even clear why my editor even worked last night, but this problem is occurring again today.

I was using `emacs-24.5-z-mac-5.8` and have also tried building from `HEAD` to see if there was any difference.

**Comment by boj** [comment]
I found this Japanese article which makes it look pretty convoluted to get working:  http://qiita.com/hiratara/items/15aba2534b16c2af8a02

**Comment by boj** [comment]
I believe I have solved the problem.

Within my `.emacs.d/elpa/ghc-#####/ghc-process.el` file line 24 I changed it from:

    (defvar ghc-command "ghc-mod")

To:

    (defvar ghc-command "stack exec -- ghc-mod")

And things seem to work now.

**Comment by hakuch** [comment]
@boj I didn't think that would work. :)

I installed the Haskell-related tools via

```
$ stack install ghc-mod
```

and by adding `~/.local/bin` to my `PATH`, I can execute them. It looks like Emacs has no problem finding the executables either. I'm still not sure why things are not working as expected.

**Comment by kvelicka** [comment]
I think this may not be the only (or perhaps the same?) issue hiding somewhere in the Haskell layer(s). 

I've recently opened #2076 about some freezes when scrolling through Haskell source files and it seems that other users have experienced something similar as well..

**Comment by codygman** [comment]
It's not a permanent fix, but I currently use this [dir-locals.el](https://github.com/commercialhaskell/stack/blob/master/.dir-locals.el) to use stack with emacs/spacemacs.

**Comment by hakuch** [comment]
@codygman I mentioned that in an earlier comment. :-)

I was able to load files into the REPL, but things like "inspect type at point" and company completion didn't seem to be working. Perhaps I've got it misconfigured?

**Comment by latkins** [comment]
I have managed to get this working by editing the haskell layer to use the settings from dir-locals.el that was posted above. I am sure there is a better way of doing it, but the settings from dir-locals.el seemed to be overwritten by the haskell layer when loading a .hs file. I haven't tested extensively, but it seems that everything works and is now using stack. This does not include autocomplete, as I have disabled that due to a bug.

**Comment by hakuch** [comment]
@latkins Can you post a snippet that I could try too, or indicate where the changes should be made in the Haskell layer?

**Comment by latkins** [comment]
In my emacs.d/contrib/!lang/haskell/packages.el I now have: 

```
      (setq haskell-process-type 'ghci
            haskell-process-path-ghci "stack"
            haskell-process-args-ghci '("ghci")
            haskell-process-args-cabal-repl '("--ghc-option=-ferror-spans")
            ...)
```

**Comment by hakuch** [comment]
Thank you kindly. I didn't notice any immediate difference, but maybe it will help others (or perhaps I still have things misconfigured).

**Comment by latkins** [comment]
I tested by loading a .hs file, then doing C-h v {variable}, to see that they were correctly set.

**Comment by codygman** [comment]
There is a [stack mode](https://github.com/commercialhaskell/stack-ide/blob/master/stack-mode/stack-mode.el) that uses [ide-backend](https://github.com/fpco/ide-backend) spacemacs might want to consider supporting. 

Maybe a different layer named haskell-stack or something.

**Comment by hakuch** [comment]
@codygam That looks neat.

**Comment by cies** [comment]
@codygman That's a very interesting question indeed. I don't know if `stack-ide` is already up to par with `ghc-mod`. I've used it in Sublime with the `stack-ide-sublime` plugin; it is really fast!

Since `stack-ide` works only for projects with `stack.yml`, but `ghc-mod` works either way (but arguably without slack build are less deterministic), I expect Spacemacs will likely end up to support both. At least until stck become ubiquitous in Haskell land.

**Comment by TakSuyu** [comment]
I put this in `dotspacemacs/config`
```lisp
(setq haskell-process-type 'stack-ghci)
```
and got the repl to work

**Comment by d12frosted** [comment]
@kvelicka about ghc-mod and 7.10, this is probably related: https://github.com/kazu-yamamoto/ghc-mod/issues/437

**Comment by boj** [comment]
kazu-yamamoto/ghc-mod#498 seems like it may fix everything (currently in master).

**Comment by fatlazycat** [comment]
Does anyone have any instructions to use stack with spacemacs ?

Thanks

**Comment by bts** [comment]
Until we have something more sustainable in place, I've had to manually replace this line https://github.com/syl20bnr/spacemacs/blob/d993a021847cde2c42865bab6afa8adbb2edda0b/contrib/!lang/haskell/packages.el#L70 with `(setq haskell-process-type 'stack-ghci` in at `~/.emacs.d/contrib/!lang/haskell/packages.el`. You'll also need the latest ghc-mod.

**Comment by d12frosted** [comment]
@fatlazycat I've created [private layer](https://github.com/d12frosted/environment/tree/master/emacs/d12-haskell) for that. Going to finish my work on it and then try to merge it into haskell layer. Currently I haven't implemented spacemacs-style keybindings (you actually can just copy and paste them from haskell layer) and flycheck support (it just breaks interactive mode for me). But it has completion both in repl and sources (using `company-ghci`). 

Example of usage:

```lisp
(setq-default dotspacemacs-configuration-layers
   '((d12-haskell :variables haskell-enable-hindent-style "chris-done"
                             haskell-enable-shm-support t
                             haskell-process-type 'stack-ghci)))
```

**Comment by d12frosted** [comment]
@bts on dev branch you can set that variable without modifying `package.el` (so soon it will be available on master as well).

**Comment by fatlazycat** [comment]
Thanks @bts and @d12frosted

**Comment by fatlazycat** [comment]
Hi, finally getting some time. 

@d12frosted trying to copy your private layer and config from spacemacs but getting 

Warning (emacs): Unable to activate package `aggressive-indent'.
Required package `names-20150125.9' is unavailable
Error (use-package): haskell-mode :config: interactive-haskell-mode is not currently registered as a minor mode

Any ideas ?

Thanks

**Comment by d12frosted** [comment]
@fatlazycat `aggressive-indent`? I think it's not from this layer, so should not harm this layer. As for `interactive-haskell-mode` - I am not sure what's wrong. Do you see this error when opening `hs` file?

**Comment by mrkgnao** [comment]
I have this set up perfectly, with `stack`. I'll try to outline my setup when I get back to my PC. It's essentially what's in [this Reddit thread](https://www.reddit.com/r/haskell/comments/3kr6m3/update_i_got_spacemacs_working_using_stack_with/).

**Comment by fatlazycat** [comment]
@d12frosted It's not obvious that it's form a diff layer and the only diff is trying to pull in your private layer. Yes when I open a hs file I get that error.

@mrkgnao thanks

**Comment by fatlazycat** [comment]
@d12frosted if I disable the private haskell layer the errors go away with aggressive-indent

**Comment by d12frosted** [comment]
@fatlazycat thanks for report, I'll investigate that. But since people found a solution for stack only setup with default layer, I advice you to use default layer instead of my private. Default layer is much more rich in functionality and `company-ghc` is better for completion than `company-ghci` (in my opinion). So really, give it a try. If you have any problems with it - just post it here or create new issue.

@mrkgnao I've tried again using `haskell` layer (and installed `ghc-mod` from soruces), but when I open any `haskell` file I see following error:

```
ghc-mod: GMEProcess "readProcessStderrChan" "/Users/d12frosted/.cabal/libexec/cabal-helper-wrapper"\
 ["/path/to/project","/path/to/project/.stack-work/dist/x86_64-osx/Cabal-1.22.4.0"] (L\
eft 1)                                                                                              
cabal-helper-wrapper: ghc: readCreateProcess: runInteractiveProcess: exec: does not exist (No such \
file or directory)                                                                                  
ghc-mod: GMEProcess "readProcessStderrChan" "/Users/d12frosted/.cabal/libexec/cabal-helper-wrapper"\
 ["/path/to/project","/path/to/project/.stack-work/dist/x86_64-osx/Cabal-1.22.4.0"] (L\
eft 1)     
```

I have `~/.cabal/libexec/cabal-helper-wrapper`. But looks like it has following usage

```
Usage: cabal-helper-wrapper ( print-appdatadir
| print-build-platform
| [--verbose]
  [--with-ghc=GHC_PATH]
  [--with-ghc-pkg=GHC_PKG_PATH]
  [--with-cabal=CABAL_PATH]
  [--with-cabal-version=VERSION]
  [--with-cabal-pkg-db=PKG_DB]
  PROJ_DIR DIST_DIR ( print-exe | [CABAL_HELPER_ARGS...] ) )
```

And from what I see (the error message), it doesn't pass `--with-ghc` and other options. You see, I don't have ghc installation other than from `stack`. So, in order to make it working - you need to make sure that  `ghc`, `ghci` and `cabal` executables are in your path. Or, you can create aliases for them - `ghc -> stack ghc` etc.

**Comment by fatlazycat** [comment]
@d12frosted thanks

@mrkgnao do you have you env checked in at all ?

**Comment by fatlazycat** [comment]
@d12frosted @mrkgnao yeah i get the same as 'readProcessStderrChan' error. I have no ghc apart from the one via stack and hence it fails.

**Comment by d12frosted** [comment]
@fatlazycat have you tried to add 'ghc' and friends location in your `PATH`? For example, `$HOME/.stack/programs/x86_64-osx/ghc-7.10.2/bin`. Or if you don't want to add this dir into your `PATH`, you could add some aliases in your shell config file.

**Comment by fatlazycat** [comment]
it's already added via /etc/paths.d entry

**Comment by d12frosted** [comment]
Hm. First of all - verify that `ghc` and `ghci` are available from your terminal (silly, but good starting point). Then start eshell in emacs and verify that `ghc` and `ghci` are available from there. Could you do it and write results?

**Comment by fatlazycat** [comment]
Ah sorry I though the changes to emacs config would allow it to use 'stack ghc' - needs a direct link to ghc ?

**Comment by d12frosted** [comment]
No, `ghc` needs a link to `stack ghc` and not vice versa. If you have installed ghc only through stack, then you don't have `ghc` in your `$PATH`. That's why you either need to add `ghc` installation dir to `$PATH` or create aliases. For example, I've configured my haskell environment using `stack` and `lts-3.4` - it uses `ghc v.7.10.2`. For me the installation dir of `ghc` is `$HOME/.stack/programs/x86_64-osx/ghc-7.10.2/bin`. So I add this path to my `$PATH`.

**Comment by fatlazycat** [comment]
Yeah now have the same issue with cabal missing that you have.

**Comment by d12frosted** [comment]
But I don't think that you need `cabal` for `stack` only setup.

**Comment by fatlazycat** [comment]
From my understanding I would agree, though would say the same about GHC - should be invoking it through stack ( i think )

**Comment by boj** [comment]
I've been successfully using spacemacs+stack for a week now with no issues.  Updated to 0.104.1 and all packages yesterday with no issues.

My .spacemacs looks like this: https://gist.github.com/boj/662e334a0beecc8f09df

My .stack/global/stack.yaml is set to use the lts-3.5 resolver, and I've installed the haskell-layer stuff using stack as usual.  However, I manually built ghc-mod from this commit https://github.com/kazu-yamamoto/ghc-mod/commit/0fde762500016d00ff3bed32d8dfca6e1b351b55

**Comment by d12frosted** [comment]
So yeah, `ghc-mod` and `stack` are playing nice together. The only trick is installing `ghc-mod` with `stack`. But now it's available in nightly resolver, so it's not that hard.

Updated readme file.

**Comment by fatlazycat** [comment]
Is it missing instructions for GHC on the path and changing haskell-process-path-ghci ?

**Comment by d12frosted** [comment]
Well, `haskell-mode` should automatically detect that you are using `stack`. Anyway, good call.

**Comment by TheBB** [comment]
The README now has ample instructions for stack users, so I am going to label this as fixed in develop.

**Comment by fatlazycat** [comment]
Sorry so you do still need ghc & the extra setting in .spacemacs until haskell-mode is updated ?

**Comment by d12frosted** [comment]
What do you mean? We don't wait for `haskell-mode` to be updated. It should already automatically infer that you are using stack. If it has any problems with it, you can help it by setting `haskell-process-type` in `.spacemacs` file.

**Comment by fatlazycat** [comment]
Sorry just confused if you actually need to gcc on your path and the extra bits in .spacemacs

**Comment by robbyoconnor** [comment]
@fatlazycat assuming your system is setup -- gcc should already be there :P

**Comment by d12frosted** [comment]
Sure, everything must be in your `$PATH`. Otherwise it will not work. You can look for example [here](https://github.com/syl20bnr/spacemacs/tree/develop/layers/+lang/haskell#setup-path).

**Comment by fatlazycat** [comment]
No sorry ghc itself for the tools in emacs. I would expect ghc to be invoked via stack and ghc not to be on the path.

**Comment by d12frosted** [comment]
Ah. If you are using `stack` then only `stack` should be on your `$PATH`. Let me update readme file.

**Comment by fatlazycat** [comment]
cool thanks

**Comment by d12frosted** [comment]
@fatlazycat done

**Comment by fatlazycat** [comment]
cheers

**Comment by d12frosted** [comment]
FYI, 

> No, ghc needs a link to stack ghc and not vice versa. If you have installed ghc only through stack, then you don't have ghc in your $PATH. That's why you either need to add ghc installation dir to $PATH or create aliases. For example, I've configured my haskell environment using stack and lts-3.4 - it uses ghc v.7.10.2. For me the installation dir of ghc is $HOME/.stack/programs/x86_64-osx/ghc-7.10.2/bin. So I add this path to my $PATH.

This is no more needed. Just checked - `haskell-mode` now handles `ghc[i]` executables properly when you are using `stack`. Just make sure that you don't have `dist` directory in project root.

**Comment by ghost355** [comment]
I nave one question - when I try to use SPC-m-h-t - I see a message like "This command need a GHCI-NG" 
But I add  this 'haskell-enable-ghci-ng-support t' and I can start ghci-ng from terminal (this in in PATH in ~/.local/bin/. Do you know how to help me? Can I use ghc-mod at the same time. Whith ghc-mod show type work well.

**Comment by d12frosted** [comment]
> Can I use ghc-mod at the same time. Whith ghc-mod show type work well.

Yes, it's possible to get `ghc-mod` functions even if you use `ghci-ng`. But you'll need to add following into your `.spacemacs` file:

```
(evil-leader/set-key-for-mode 'haskell-mode
        "mhi"  'ghc-show-info
        "mht"  'ghc-show-type)
```

As for initial problem. AFAIK, you need active REPL in order to make `m h t` and `m h i` functions from `ghci-ng` working. Does it make sense?

**Comment by ghost355** [comment]
If I remove (haskell-enable-ghci-ng-support t) and remove (haskell-enable-ghc-mod-support t) from .spacemacs everything work fine. M-x haskell-session-change and SPC-m-s-и (haskell-process-load-or-reload) and SPC-m-h-t show me type and T insert type signature. I don't know which of both ghc-mod or ghci-ng working for it )))

**Comment by d12frosted** [comment]
@ghost355 `haskell-enable-ghc-mod-support` is `t` by default, so what you get is from `ghc-mod`. If you want to use `ghci-ng` and still have bindings from `ghc-mod`, you should use snipped I posted earlier.

As for `ghci-ng` bindings - you need to start interactive session in order to make them working. So you should set `haskell-enable-ghci-ng-support` to `t` - before starting interactive session (REPL) pressing `spc m h t` should say that it requires `ghci-ng`. Then try starting interactive session and press that keybinding again - should be working. If it's not - let me know :smile:

**Comment by ghost355** [comment]
Thanks. My report:
1. Add `haskell-enable-ghci-ng-support t` (remove `haskell-enable-ghc-mod-support t`).Write and start Spacemacs
2. Load Main.hs file which was created by `stack new name_project simple`
3. SPC-m-h-t Message "No session... try `M-x haskell-session-change`
4.  `M-x haskell-session-change`? If I run it I see new window with REPL 
5. OK, after 4) I and pressing SPC-m-h-t I see "This command requires GHCI-ng... "

**Comment by d12frosted** [comment]
I'll take a look into `ghci-ng` problem on weekends. Looks like something is not working as good as expected with it.

**Comment by d12frosted** [comment]
Checked out. There are some caveats with using `ghci-ng` in the stack-driven project. Described them in #3544.

**Comment by ghost355** [comment]
Thanks, @d12frosted. 
What benefits  of ghci-ng?

**Comment by fatlazycat** [comment]
Personally I use ghci-ng for type info on a selection. Find it invaluable.

**Comment by d12frosted** [comment]
@ghost355 find usages, go-to-definition, type of selection (as @fatlazycat says). But I don't like that it requires manual REPL initialisation. What I like about `ghc-mod` - it starts automatically in bg.

**Comment by ghost355** [comment]
Ghc-mod doing its functions too, isn't it? Or I mistaken?

**Comment by d12frosted** [comment]
@ghost355 
`ghc-mod` is doing it's functions, but it doesn't have go-to-definition, find usages and type of selection. Go to definition in haskell layer is done via TAGS. As for type of selection, `ghc-mod` supports only type of symbol, but `ghci-ng` supports asking type of `putStrLn "Hello"` (which will return `IO ()`).

**Comment by ghost355** [comment]
And one more question - is it possible start Haskell session automatically when I open .hs file? Like M-x Haskell-change-session. Do I need using hook oh open Haskell files for it?

**Comment by d12frosted** [comment]
Well yes, it's possible - via hooks. The problem is that `ghc-mod` operates on project, while `ghci-ng` operates on parts of code (usually buffers). So you'll have to use `haskell-interactive-load-or-reload` every time you visit (via find, or just buffer switch) `.hs` file. Honestly, I don't like this idea - too laggy should it be.

**Comment by ghost355** [comment]
No, you understood right ;)
I understand. Ghci-ng better, but need extra motion for activation )))

**Comment by d12frosted** [comment]
> I understand. Ghci-ng better, but need extra motion for activation )))

Yeah, it requires more motion. But not that `ghci-ng` is better - it has functionality that `ghc-mod` doesn't have. And vice versa, `ghc-mod` has functionality that `ghci-ng` lacks. Just to name most important for me - autocompletion (not that it's as important as in other languages, but it makes my life easier). 

BTW, if you are interested in ide tooling for haskell, checkout [haskell-ide](https://github.com/haskell/haskell-ide) project. Seems like they want to combine existing tools into one to avoid these problems with thousands tools.

**Comment by ghost355** [comment]
Yes, haskell-ide is good idea, but they still have a lot of work for it

**Comment by d12frosted** [comment]
@ghost355 indeed, a lot of - AFAIK, they want to test architecture with some simple features, and then start call-to-arms :smile:

**Comment by ghost355** [comment]
We'll see )))
поживем - увидим!

**Comment by ghost355** [comment]
thanks one more @d12frosted ))) Both ghci-ng and ghc-mod functions works together well, I think.

**Comment by d12frosted** [comment]
@ghost355 glad to hear - if anything becomes funky - just open new issue or ping me on ![Gitter](https://badges.gitter.im/Join Chat.svg) :smile:

**Comment by ghost355** [comment]
sorry, but I don't open gitter link :(

**Comment by robbyoconnor** [comment]
http://gitter.im/syl20bnr/spacemacs <- @ghost355

**Comment by d12frosted** [comment]
Oops, wrong link - my bad :smile: thanks, @robbyoconnor !

**Comment by x-ji** [comment]
For some reason, even if I set `haskell-process-type` to `stack-ghci` manually, Emacs would still say it can't find `ghci`. I had to manually add the directory `/home/jx/.stack/programs/x86_64-linux/ghc-7.10.2/bin` for the REPL to start up normally. I'm not sure what happened here.
